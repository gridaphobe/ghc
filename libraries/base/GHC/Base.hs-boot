{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Base where

import GHC.Types

(++) :: [a] -> [a] -> [a]

type String = [Char]
