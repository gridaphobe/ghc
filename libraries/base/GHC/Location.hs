{-# LANGUAGE RecordWildCards #-}
module GHC.Location
  ( CallStack
  , getCallStack

  , SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol

  -- * Pretty printing
  , showCallStack
  , showSrcLoc
  ) where

--------------------------------------------------------------------------------
-- Source Locations
--------------------------------------------------------------------------------

-- | A CallStack is comprised of at least one 'SrcLoc'.
data CallStack = CallStack { getCallStack :: [SrcLoc] }
  deriving (Show, Eq)

showCallStack :: CallStack -> String
showCallStack (CallStack (root:rest))
  = unlines (showSrcLoc root : map (indent . showSrcLoc) rest)
  where
  indent l = "  " ++ l
showCallStack _ = error "CallStack cannot be empty!"

-- | A single location in the source code.
data SrcLoc = SrcLoc
  { srcLocPackage   :: String
  , srcLocModule    :: String
  , srcLocFile      :: String
  , srcLocStartLine :: Int
  , srcLocStartCol  :: Int
  , srcLocEndLine   :: Int
  , srcLocEndCol    :: Int
  } deriving (Show, Eq)

showSrcLoc :: SrcLoc -> String
showSrcLoc SrcLoc {..}
  = concat [ srcLocFile, ":"
           , show srcLocStartLine, ":"
           , show srcLocStartCol, " in "
           , srcLocPackage, ":", srcLocModule
           ]
