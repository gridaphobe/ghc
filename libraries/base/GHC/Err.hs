{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude, MagicHash #-}
{-# OPTIONS_HADDOCK hide #-}

-----------------------------------------------------------------------------
-- |
-- Module      :  GHC.Err
-- Copyright   :  (c) The University of Glasgow, 1994-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- The "GHC.Err" module defines the code for the wired-in error functions,
-- which have a special type in the compiler (with \"open tyvars\").
--
-- We cannot define these functions in a module where they might be used
-- (e.g., "GHC.Base"), because the magical wired-in type will get confused
-- with what the typechecker figures out.
--
-----------------------------------------------------------------------------

module GHC.Err( error ) where
import GHC.CString ()
import GHC.Types
import GHC.Prim
import GHC.Integer ()   -- Make sure Integer is compiled first
                        -- because GHC depends on it in a wired-in way
                        -- so the build system doesn't see the dependency
import {-# SOURCE #-} GHC.Exception( errorCallException )

-- | 'error' stops execution and displays an error message.
error :: [Char] -> CallStack -> a
error s stk = raise# (errorCallException s stk)
