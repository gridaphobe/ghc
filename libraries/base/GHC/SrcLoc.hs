{-# LANGUAGE NoImplicitPrelude, RecordWildCards #-}

-- | @since 4.8.2.0
module GHC.SrcLoc
  ( SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol

  -- * Pretty printing
  , showSrcLoc
  ) where

import GHC.Base
import GHC.Show

-- | Pretty print 'SrcLoc'
--
-- @since 4.8.2.0
showSrcLoc :: SrcLoc -> [Char]
showSrcLoc SrcLoc {..}
  = foldr (++) ""
      [ srcLocFile, ":"
      , show srcLocStartLine, ":"
      , show srcLocStartCol, " in "
      , srcLocPackage, ":", srcLocModule
      ]

