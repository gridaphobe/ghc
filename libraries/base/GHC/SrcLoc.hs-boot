{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.SrcLoc
  ( SrcLoc
  , srcLocPackage
  , srcLocModule
  , srcLocFile
  , srcLocStartLine
  , srcLocStartCol
  , srcLocEndLine
  , srcLocEndCol
  ) where

import GHC.Types( Char, Int )

-- | A single location in the source code.
data SrcLoc = SrcLoc
  { srcLocPackage   :: [Char]
  , srcLocModule    :: [Char]
  , srcLocFile      :: [Char]
  , srcLocStartLine :: Int
  , srcLocStartCol  :: Int
  , srcLocEndLine   :: Int
  , srcLocEndCol    :: Int
  }
