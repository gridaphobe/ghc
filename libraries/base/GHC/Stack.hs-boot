-- This SOURCE-imported hs-boot module cuts a dependency loop:
--
--          GHC.Stack
-- imports  GHC.Base
-- imports  GHC.Err
-- imports  {-# SOURCE #-} GHC.Stack

{-# LANGUAGE Trustworthy #-}
{-# LANGUAGE NoImplicitPrelude #-}

module GHC.Stack ( CallStack, showCallStack ) where

import {-# SOURCE #-} GHC.SrcLoc ( SrcLoc )
import GHC.Types ( Char )

data CallStack = CallStack { getCallStack :: [([Char], SrcLoc)] }

showCallStack :: CallStack -> [Char]
