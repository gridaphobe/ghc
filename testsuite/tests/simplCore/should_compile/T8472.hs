{-# LANGUAGE MagicHash #-}
{-# LANGUAGE CPP #-}

module T8472(t) where

#if ADDR
import GHC.Exts(Ptr(..), Addr#)
import Foreign.Ptr
#else
import GHC.Exts(Int(..), Int#)
#endif

type Fun = (Int, Int -> IO ())

gve :: Int -> (Int -> IO ()) -> Fun
gve n f = (n, f . (n+))
{-# INLINE gve #-}

#if ADDR
fun :: Addr# -> Int -> IO ()
fun addr = \n -> print (Ptr addr `plusPtr` n)
#else
fun :: Int# -> Int -> IO ()
fun x = \n -> print (I# x + n)
#endif
{-# INLINE fun #-}

comb :: Fun -> Fun -> Fun
comb (n, f) (m, g) = (n + m, \x -> f x >> g x)

t :: Fun -> [Fun]
t n =
  [ n
  `comb`
#if ADDR
  gve 3 (fun "foo"#)
  `comb`
  gve 3 (fun "bar"#)
#else
  gve 3 (fun 9897114#)
  `comb`
  gve 3 (fun 9897122#)
#endif
  ]
