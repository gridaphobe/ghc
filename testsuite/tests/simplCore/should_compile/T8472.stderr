
==================== Tidy Core ====================
Result size of Tidy Core = {terms: 91, types: 103, coercions: 7}

-- RHS size: {terms: 1, types: 0, coercions: 0}
$trModule4
$trModule4 = "main"#

-- RHS size: {terms: 2, types: 0, coercions: 0}
$trModule3
$trModule3 = TrNameS $trModule4

-- RHS size: {terms: 1, types: 0, coercions: 0}
$trModule2
$trModule2 = "T8472"#

-- RHS size: {terms: 2, types: 0, coercions: 0}
$trModule1
$trModule1 = TrNameS $trModule2

-- RHS size: {terms: 3, types: 0, coercions: 0}
$trModule
$trModule = Module $trModule3 $trModule1

-- RHS size: {terms: 1, types: 0, coercions: 0}
t2
t2 = "foo"#

-- RHS size: {terms: 1, types: 0, coercions: 0}
t1
t1 = "bar"#

-- RHS size: {terms: 63, types: 75, coercions: 7}
$wt
$wt =
  \ w_s4AR ->
    (# case w_s4AR of { (n_a3kF, f_a3kG) ->
       (case n_a3kF of { I# x_a3Tx -> I# (+# (+# x_a3Tx 3#) 3#) },
        (\ x_a3kJ s_a4k6 ->
           case ((f_a3kG x_a3kJ) `cast` ...) s_a4k6 of
           { (# ipv_a4k9, ipv1_a4ka #) ->
           case hPutStr2
                  stdout
                  (case x_a3kJ of { I# y_a3TB ->
                   case $w$cshowsPrec (plusAddr# t2 (+# 3# y_a3TB)) [] of
                   { (# ww3_a4jW, ww4_a4jX #) ->
                   : ww3_a4jW ww4_a4jX
                   }
                   })
                  True
                  ipv_a4k9
           of
           { (# ipv2_X4kG, ipv3_X4kI #) ->
           hPutStr2
             stdout
             (case x_a3kJ of { I# y_a3TB ->
              case $w$cshowsPrec (plusAddr# t1 (+# 3# y_a3TB)) [] of
              { (# ww3_a4jW, ww4_a4jX #) ->
              : ww3_a4jW ww4_a4jX
              }
              })
             True
             ipv2_X4kG
           }
           })
        `cast` ...)
       },
       [] #)

-- RHS size: {terms: 8, types: 11, coercions: 0}
t
t =
  \ w_s4AR ->
    case $wt w_s4AR of { (# ww1_s4AW, ww2_s4AX #) ->
    : ww1_s4AW ww2_s4AX
    }



